lib/aiken/collection/pairs.ak

```diff
@@ -572,6 +573,106 @@ test map_2() {
   map(fixture, with: fn(_, v) { v + 1 }) == [Pair("a", 2), Pair("b", 3)]
 }
 
+/// Merge a value into the `Pairs` at a given key. If the key already exists,
+/// its value is merged by the merging function.
+///
+/// ```aiken
+/// use aiken/builtin
+/// use aiken/primitive/bytearray
+///
+/// let compare_un_b_data = fn(l, r) {
+///   bytearray.compare(l |> builtin.un_b_data, r |> builtin.un_b_data)
+/// }
+///
+/// let result =
+///   []
+///     |> pairs.merge(key: "foo" |> builtin.b_data, value: 1, compare: compare_un_b_data, merging: builtin.add_integer)
+///     |> pairs.merge(key: "bar" |> builtin.b_data, value: 2, compare: compare_un_b_data, merging: builtin.add_integer)
+///     |> pairs.merge(key: "foo" |> builtin.b_data, value: 3, compare: compare_un_b_data, merging: builtin.add_integer)
+///
+/// result == [Pair("bar" |> builtin.b_data, 2), Pair("foo" |> builtin.b_data, 4)]
+/// ```
+///
+/// Notice that the key is of type `Data`. The function works with any key type, not just ByteArray.
+/// Unlike `dict.from_pairs() |> dict.union_with() |> dict.to_pairs()`
+pub fn merge(
+  self: Pairs<key, value>,
+  key k: key,
+  value v: value,
+  compare: fn(key, key) -> Ordering,
+  merging: fn(value, value) -> value,
+) -> Pairs<key, value> {
+  when self is {
+    [] ->
+      [Pair(k, v)]
+
+    [Pair(k2, v2), ..rest] ->
+      when compare(k, k2) is {
+        Less ->
+          [Pair(k, v), ..self]
```

Comment:
That is unfortunately not possible in this context. `Pairs` do not make any assumption about the ordering of keys; It's fundamentally just a list. So we cannot short-circuit merging as soon as we meet a higher key; we have to traverse it in full; always.